mesonbuild/modules/simd.py

```diff
@@ -13,76 +13,109 @@
 # limitations under the License.
 from __future__ import annotations
 
-from .. import mesonlib, compilers, mlog
+import typing as T
+
+from .. import mesonlib, mlog
 from .. import build
+from ..compilers import Compiler
+from ..interpreter.type_checking import BT_SOURCES_KW, STATIC_LIB_KWS
+from ..interpreterbase.decorators import KwargInfo, typed_pos_args, typed_kwargs
 
 from . import ExtensionModule, ModuleInfo
 
+if T.TYPE_CHECKING:
+    from . import ModuleState
+    from ..interpreter import Interpreter, kwargs as kwtypes
+    from ..interpreter.type_checking import SourcesVarargsType
+
+    class CheckKw(kwtypes.StaticLibrary):
+
+        compiler: Compiler
+        mmx: SourcesVarargsType
+        sse: SourcesVarargsType
+        sse2: SourcesVarargsType
+        sse3: SourcesVarargsType
+        ssse3: SourcesVarargsType
+        sse41: SourcesVarargsType
+        sse42: SourcesVarargsType
+        avx: SourcesVarargsType
+        avx2: SourcesVarargsType
+        neon: SourcesVarargsType
+
+
+# FIXME add Altivec and AVX512.
+ISETS = (
+    'mmx',
+    'sse',
+    'sse2',
+    'sse3',
+    'ssse3',
+    'sse41',
+    'sse42',
+    'avx',
+    'avx2',
+    'neon',
+)
+
+
 class SimdModule(ExtensionModule):
 
     INFO = ModuleInfo('SIMD', '0.42.0', unstable=True)
 
-    def __init__(self, interpreter):
+    def __init__(self, interpreter: Interpreter):
         super().__init__(interpreter)
-        # FIXME add Altivec and AVX512.
-        self.isets = ('mmx',
-                      'sse',
-                      'sse2',
-                      'sse3',
-                      'ssse3',
-                      'sse41',
-                      'sse42',
-                      'avx',
-                      'avx2',
-                      'neon',
-                      )
         self.methods.update({
             'check': self.check,
         })
 
-    def check(self, state, args, kwargs):
-        result = []
-        if len(args) != 1:
-            raise mesonlib.MesonException('Check requires one argument, a name prefix for checks.')
-        prefix = args[0]
-        if not isinstance(prefix, str):
-            raise mesonlib.MesonException('Argument must be a string.')
-        if 'compiler' not in kwargs:
-            raise mesonlib.MesonException('Must specify compiler keyword')
+    @typed_pos_args('simd.check', str)
+    @typed_kwargs('simd.check',
+                  KwargInfo('compiler', Compiler, required=True),
+                  *[BT_SOURCES_KW.evolve(name=iset) for iset in ISETS],
+                  *[a for a in STATIC_LIB_KWS if a.name != 'sources'],
+                  allow_unknown=True) # Because we also accept STATIC_LIB_KWS, but build targets have not been completely ported to typed_pos_args/typed_kwargs.
```

Comment:
I *think* we need to add `permitted_kwargs(build.known_stlib_kwargs)` here since we are removing the invocation of interpreter.func_static_lib and using interpreter.build_target directly.

Other than that, this looks good.

Reply:
Right, clearly I meant that even if I didn't actually know I meant that. :D